model/Task/AbstractTask.php

```diff
@@ -0,0 +1,223 @@
+            throw new \InvalidArgumentException(sprintf(
+                'Expected a string, array, or Traversable argument in first position; received "%s"',
+                (is_object($spec) ? get_class($spec) : gettype($spec))
+            ));
+        }
+
+        foreach ($spec as $key => $value) {
+            $this->metadata[$key] = $value;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Retrieve a single metadata as specified by key
+     *
+     * @param  string $key
+     * @param  null|mixed $default
+     * @throws \InvalidArgumentException
+     * @return mixed
+     */
+    public function getMetadata($key, $default = null)
+    {
+        if (!is_string($key)) {
+            throw new \InvalidArgumentException('Non-string argument provided as a metadata key');
+        }
+
+        if (array_key_exists($key, $this->metadata)) {
+            return $this->metadata[$key];
+        }
+
+        return $default;
+    }
+
+    /**
+     * Set task parameter
+     *
+     * @param  string|array|\Traversable $spec
+     * @param  mixed $value
+     * @throws \InvalidArgumentException
+     * @return TaskInterface
+     */
+    public function setParameter($spec, $value = null)
+    {
+        if (is_string($spec)) {
+            $this->parameters[$spec] = $value;
+            return $this;
+        }
+
+        if (!is_array($spec) && !$spec instanceof \Traversable) {
+            throw new \InvalidArgumentException(sprintf(
+                'Expected a string, array, or Traversable argument in first position; received "%s"',
+                (is_object($spec) ? get_class($spec) : gettype($spec))
+            ));
+        }
+
+        foreach ($spec as $key => $value) {
+            $this->parameters[$key] = $value;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Retrieve a single parameter as specified by key
+     *
+     * @param  string $key
+     * @param  null|mixed $default
+     * @throws \InvalidArgumentException
+     * @return mixed
+     */
+    public function getParameter($key, $default = null)
+    {
+        if (!is_string($key)) {
+            throw new \InvalidArgumentException('Non-string argument provided as a parameter key');
+        }
+
+        if (array_key_exists($key, $this->parameters)) {
+            return $this->parameters[$key];
+        }
+
+        return $default;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * @param \DateTime $dateTime
+     * @return TaskInterface
+     */
+    public function setCreatedAt(\DateTime $dateTime)
+    {
+        $this->setMetadata('__created_at__', $dateTime);
+
+        return $this;
+    }
+
+    /**
+     * @return \DateTime
+     */
+    public function getCreatedAt()
+    {
+        if (($date = $this->getMetadata('__created_at__')) && is_string($date)) {
+            $date = (new \DateTime($date));
+        }
+
```

Comment:
So we are doing also an deserialize of the date each time we call getCreatedAt(). If we don't have a common place where we deserialize the object, then after deserializing call $this->setCreatedAt($date).